To enforce this:
```diff
diff --git a/homeassistant/data_entry_flow.py b/homeassistant/data_entry_flow.py
index cb66dbf521b..79579d83259 100644
--- a/homeassistant/data_entry_flow.py
+++ b/homeassistant/data_entry_flow.py
@@ -572,6 +572,24 @@ class FlowHandler:
             schema[new_key] = val
         return vol.Schema(schema)

+    def check_step(self, step_id: str) -> None:
+        import inspect
+
+        curframe = inspect.currentframe()
+        calframe = inspect.getouterframes(curframe, 2)
+        idx = 2
+        while idx < len(calframe):
+            calmodule = calframe[idx][1]
+            if calmodule.endswith("schema_config_entry_flow.py"):
+                return
+                pass
+            calname = calframe[idx][3]
+            if calname.startswith("async_step_"):
+                break
+            idx += 1
+        if not calname == f"async_step_{step_id}":
+            raise ValueError(f"{calname} != async_step_{step_id}")
+
     @callback
     def async_show_form(
         self,
@@ -599,6 +617,8 @@ class FlowHandler:
         )
         if step_id is not None:
             flow_result["step_id"] = step_id
+            self.check_step(step_id)
+
         return flow_result

     @callback
@@ -659,6 +679,7 @@ class FlowHandler:
         )
         if step_id is not None:
             flow_result["step_id"] = step_id
+            self.check_step(step_id)
         return flow_result

     @callback
@@ -694,6 +715,7 @@ class FlowHandler:
         )
         if step_id is not None:
             flow_result["step_id"] = step_id
+            self.check_step(step_id)
         return flow_result

     @callback
@@ -729,6 +751,7 @@ class FlowHandler:
         )
         if step_id is not None:
             flow_result["step_id"] = step_id
+            self.check_step(step_id)
         return flow_result

     @callback
```

The following tests fail when enforcing this in tests:
```
FAILED tests/components/airzone_cloud/test_config_flow.py::test_form - ValueError: async_step_inst_pick != async_step_user
FAILED tests/components/airzone_cloud/test_config_flow.py::test_installations_list_error - ValueError: async_step_inst_pick != async_step_user
FAILED tests/components/elgato/test_config_flow.py::test_full_zeroconf_flow_implementation - ValueError: async_step_zeroconf != async_step_zeroconf_confirm
FAILED tests/components/heos/test_config_flow.py::test_discovery_shows_create_form - ValueError: async_step_ssdp != async_step_user
FAILED tests/components/heos/test_config_flow.py::test_discovery_sets_the_unique_id - ValueError: async_step_ssdp != async_step_user
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_works[True-True] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_works[True-False] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_works[False-True] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_works[False-False] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_abort_duplicate_flow - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_does_not_ignore_non_homekit - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_invalid_config_entry - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_abort_errors_on_start[AccessoryNotFoundError-accessory_not_found_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_abort_errors_on_start[UnavailableError-already_paired] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_try_later_errors_on_start[BusyError-busy_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_try_later_errors_on_start[MaxTriesError-max_tries_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_try_later_errors_on_start[IndexError-protocol_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_start[KeyError-pairing_failed] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_abort_errors_on_finish[AccessoryNotFoundError-accessory_not_found_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_finish[MalformedPinError-authentication_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_finish[MaxPeersError-max_peers_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_finish[AuthenticationError-authentication_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_finish[UnknownError-unknown_error] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_form_errors_on_finish[KeyError-pairing_failed] - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_pair_unknown_errors - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_unignore_works - ValueError: async_step_unignore != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_discovery_dismiss_existing_flow_on_paired - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_mdns_update_to_paired_during_pairing - ValueError: async_step_zeroconf != async_step_pair
FAILED tests/components/homekit_controller/test_config_flow.py::test_bluetooth_valid_device_discovery_unpaired - ValueError: async_step_bluetooth != async_step_pair
FAILED tests/components/huawei_lte/test_config_flow.py::test_ssdp[requests_mock_request_kwargs0-upnp_data0-expected_result0] - ValueError: async_step_ssdp != async_step_user
FAILED tests/components/huawei_lte/test_config_flow.py::test_ssdp[requests_mock_request_kwargs1-upnp_data1-expected_result1] - ValueError: async_step_ssdp != async_step_user
FAILED tests/components/hyperion/test_config_flow.py::test_auth_create_token_approval_declined - ValueError: async_step_create_token != async_step_create_token_external
FAILED tests/components/hyperion/test_config_flow.py::test_auth_create_token_approval_declined_task_canceled - ValueError: async_step_create_token != async_step_create_token_external
FAILED tests/components/hyperion/test_config_flow.py::test_auth_create_token_when_issued_token_fails - ValueError: async_step_create_token != async_step_create_token_external
FAILED tests/components/hyperion/test_config_flow.py::test_auth_create_token_success - ValueError: async_step_create_token != async_step_create_token_external
FAILED tests/components/hyperion/test_config_flow.py::test_auth_create_token_success_but_login_fail - ValueError: async_step_create_token != async_step_create_token_external
FAILED tests/components/improv_ble/test_config_flow.py::test_provision_retry[exc0-unable_to_connect] - ValueError: <module> != async_step_provision
FAILED tests/components/improv_ble/test_config_flow.py::test_provision_retry[exc0-unable_to_connect] - ValueError: <module> != async_step_provision
FAILED tests/components/nuki/test_config_flow.py::test_form - ValueError: async_step_validate != async_step_user
FAILED tests/components/nuki/test_config_flow.py::test_form_invalid_auth - ValueError: async_step_validate != async_step_user
FAILED tests/components/nuki/test_config_flow.py::test_form_cannot_connect - ValueError: async_step_validate != async_step_user
FAILED tests/components/nuki/test_config_flow.py::test_form_unknown_exception - ValueError: async_step_validate != async_step_user
FAILED tests/components/nuki/test_config_flow.py::test_form_already_configured - ValueError: async_step_validate != async_step_user
FAILED tests/components/nuki/test_config_flow.py::test_dhcp_flow - ValueError: async_step_validate != async_step_user
FAILED tests/components/peco/test_config_flow.py::test_meter_value_error - ValueError: async_step_finish_smart_meter != async_step_user
FAILED tests/components/peco/test_config_flow.py::test_unresponsive_meter_error - ValueError: async_step_finish_smart_meter != async_step_user
FAILED tests/components/peco/test_config_flow.py::test_meter_http_error - ValueError: async_step_finish_smart_meter != async_step_user
FAILED tests/components/plex/test_config_flow.py::test_bad_credentials - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_bad_hostname - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_unknown_exception - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_no_servers_found - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_single_available_server - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_multiple_servers_with_selection - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_adding_last_unconfigured_server - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_all_available_servers_configured - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_external_timed_out - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_callback_view - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_manual_config - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_reauth - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/plex/test_config_flow.py::test_reauth_multiple_servers_available - ValueError: async_step_plex_website_auth != async_step_obtain_token
FAILED tests/components/starline/test_config_flow.py::test_flow_works - AssertionError: assert 'auth_app' == 'auth_user'
FAILED tests/components/switcher_kis/test_config_flow.py::test_user_setup[mock_bridge0] - ValueError: async_step_user != async_step_confirm
FAILED tests/components/switcher_kis/test_config_flow.py::test_user_setup_abort_no_devices_found - ValueError: async_step_user != async_step_confirm
FAILED tests/components/tedee/test_config_flow.py::test_reauth_flow - ValueError: async_step_reauth != async_step_user
FAILED tests/components/vilfo/test_config_flow.py::test_form_cannot_connect - AssertionError: assert {'base': 'unknown'} == {'base': 'cannot_connect'}
FAILED tests/components/vizio/test_config_flow.py::test_user_start_pairing_failure - ValueError: async_step_pair_tv != async_step_user
```

--- a/homeassistant/data_entry_flow.py
+++ b/homeassistant/data_entry_flow.py
@@ -575,25 +575,30 @@ class FlowHandler:
     def async_show_form(
         self,
         *,
-        step_id: str,
+        step_id: str | None = None,
         data_schema: vol.Schema | None = None,
         errors: dict[str, str] | None = None,
         description_placeholders: Mapping[str, str | None] | None = None,
         last_step: bool | None = None,
         preview: str | None = None,
     ) -> FlowResult:
-        """Return the definition of a form to gather user input."""
-        return FlowResult(
+        """Return the definition of a form to gather user input.
+
+        The step_id parameter is deprecated and will be removed in a future release.
+        """
+        flow_result = FlowResult(
             type=FlowResultType.FORM,
             flow_id=self.flow_id,
             handler=self.handler,
-            step_id=step_id,
             data_schema=data_schema,
             errors=errors,
             description_placeholders=description_placeholders,
             last_step=last_step,  # Display next or submit button in frontend
             preview=preview,  # Display preview component in frontend
         )
+        if step_id is not None:
+            flow_result["step_id"] = step_id
+        return flow_result
 
     @callback
     def async_create_entry(
@@ -636,19 +641,24 @@ class FlowHandler:
     def async_external_step(
         self,
         *,
-        step_id: str,
+        step_id: str | None = None,
         url: str,
         description_placeholders: Mapping[str, str] | None = None,
     ) -> FlowResult:
-        """Return the definition of an external step for the user to take."""
-        return FlowResult(
+        """Return the definition of an external step for the user to take.
+
+        The step_id parameter is deprecated and will be removed in a future release.
+        """
+        flow_result = FlowResult(
             type=FlowResultType.EXTERNAL_STEP,
             flow_id=self.flow_id,
             handler=self.handler,
-            step_id=step_id,
             url=url,
             description_placeholders=description_placeholders,
         )
+        if step_id is not None:
+            flow_result["step_id"] = step_id
+        return flow_result
 
     @callback
     def async_external_step_done(self, *, next_step_id: str) -> FlowResult:
@@ -669,7 +679,10 @@ class FlowHandler:
         description_placeholders: Mapping[str, str] | None = None,
         progress_task: asyncio.Task[Any] | None = None,
     ) -> FlowResult:
-        """Show a progress message to the user, without user input allowed."""
+        """Show a progress message to the user, without user input allowed.
+
+        The step_id parameter is deprecated and will be removed in a future release.
+        """
         if progress_task is None and not self.__no_progress_task_reported:
             self.__no_progress_task_reported = True
             cls = self.__class__
@@ -685,7 +698,7 @@ class FlowHandler:
                 report_issue,
             )
 
-        result = FlowResult(
+        flow_result = FlowResult(
             type=FlowResultType.SHOW_PROGRESS,
             flow_id=self.flow_id,
             handler=self.handler,
@@ -694,8 +707,8 @@ class FlowHandler:
             progress_task=progress_task,
         )
         if step_id is not None:
-            result["step_id"] = step_id
-        return result
+            flow_result["step_id"] = step_id
+        return flow_result
 
     @callback
     def async_show_progress_done(self, *, next_step_id: str) -> FlowResult:
@@ -711,23 +724,26 @@ class FlowHandler:
     def async_show_menu(
         self,
         *,
-        step_id: str,
+        step_id: str | None = None,
         menu_options: list[str] | dict[str, str],
         description_placeholders: Mapping[str, str] | None = None,
     ) -> FlowResult:
         """Show a navigation menu to the user.
 
         Options dict maps step_id => i18n label
+        The step_id parameter is deprecated and will be removed in a future release.
         """
-        return FlowResult(
+        flow_result = FlowResult(
             type=FlowResultType.MENU,
             flow_id=self.flow_id,
             handler=self.handler,
-            step_id=step_id,
             data_schema=vol.Schema({"next_step_id": vol.In(menu_options)}),
             menu_options=menu_options,
             description_placeholders=description_placeholders,
         )
+        if step_id is not None:
+            flow_result["step_id"] = step_id
+        return flow_result
 
     @callback
     def async_remove(self) -> None:

--- a/homeassistant/data_entry_flow.py
+++ b/homeassistant/data_entry_flow.py
@@ -76,6 +76,9 @@ FLOW_NOT_COMPLETE_STEPS = {
 }
 
 STEP_ID_OPTIONAL_STEPS = {
+    FlowResultType.EXTERNAL_STEP,
+    FlowResultType.FORM,
+    FlowResultType.MENU,
     FlowResultType.SHOW_PROGRESS,
 }
 

